lineapy/plugins/taskgen.py

```diff
@@ -0,0 +1,221 @@
+        return get_artifact_task_definitions(
+            artifact_collection, pipeline_name
+        )
+    else:
+        raise ValueError(
+            f"Task breakdown granularity {task_breakdown} is not currently supported."
+        )
+
+
+def get_artifact_task_definitions(
+    artifact_collection: ArtifactCollection, pipeline_name: str
+) -> Dict[str, TaskDefinition]:
+    """
+    get_artifact_task_definitions returns a task definition for each artifact the pipeline produces.
+    This may include tasks that produce common variables that were not initially defined as artifacts.
+    """
+    task_definitions: Dict[str, TaskDefinition] = dict()
+    unused_input_parameters = set(artifact_collection.input_parameters)
+
+    for session_artifacts in artifact_collection.sort_session_artifacts():
+        for nc in session_artifacts.usercode_nodecollections:
+            all_input_variables = sorted(list(nc.input_variables))
+            artifact_user_input_variables = [
+                var
+                for var in all_input_variables
+                if var in unused_input_parameters
+            ]
+            session_input_parameters_spec = (
+                BaseSessionWriter().get_session_input_parameters_spec(
+                    session_artifacts
+                )
+            )
+            user_input_var_typing_block = [
+                f"{var} = {session_input_parameters_spec[var].value_type}({var})"
+                for var in artifact_user_input_variables
+            ]
+            unused_input_parameters.difference_update(
+                set(artifact_user_input_variables)
+            )
+            loaded_input_vars = [
+                var
+                for var in all_input_variables
+                if var not in artifact_user_input_variables
+            ]
+            function_call_block = (
+                BaseSessionWriter().get_session_artifact_function_call_block(
+                    nc,
+                    source_module=f"{pipeline_name}_module",
+                )
+            )
+
+            task_def: TaskDefinition = TaskDefinition(
+                function_name=nc.safename,
+                user_input_variables=artifact_user_input_variables,
+                loaded_input_variables=loaded_input_vars,
+                typing_blocks=user_input_var_typing_block,
+                call_block=function_call_block,
+                return_vars=nc.return_variables,
+                pipeline_name=pipeline_name,
+            )
+            task_definitions[nc.safename] = task_def
+
+    return task_definitions
+
+
+def get_session_task_definition(
+    artifact_collection: ArtifactCollection, pipeline_name: str
+) -> Dict[str, TaskDefinition]:
+    """
+    get_session_task_definition returns a task definition for each session in the pipeline.
+    """
+    task_definitions: Dict[str, TaskDefinition] = dict()
+
+    for session_artifacts in artifact_collection.sort_session_artifacts():
```

Comment:
Same feedback as in https://github.com/LineaLabs/lineapy/pull/831/files#r1011015900